lib/catalog.ex

```diff
@@ -0,0 +1,121 @@
+defmodule Catalog do
+  @moduledoc """
+  A listing of all documents in our collection of buckets. The buckets are specified in the application configuration. Each document is associated with a person, and can be searched by that person's last name, first name, and/or date of birth.
+  """
+  use GenServer
+
+  alias Catalog.{Bucket, Document}
+
+  @type t :: %__MODULE__{
+          ets: :ets.tid(),
+          loaded?: boolean()
+        }
+
+  @enforce_keys [:ets, :loaded?]
+
+  defstruct ets: nil,
+            loaded?: false
+
+  # Client
+
+  @spec start_link(keyword()) :: GenServer.on_start()
+  def start_link(opts) do
+    name = Keyword.get(opts, :name, __MODULE__)
+
+    GenServer.start_link(__MODULE__, opts, name: name)
+  end
+
+  @spec lookup(String.t() | nil, String.t() | nil, Date.t() | nil) ::
+          {:ok, [Document.t()]} | {:error, String.t()}
+  @spec lookup(GenServer.server(), String.t() | nil, String.t() | nil, Date.t() | nil) ::
+          {:ok, [Document.t()]} | {:error, String.t()}
+  def lookup(pid \\ __MODULE__, last_name, first_name, date_of_birth)
+
+  def lookup(_, nil, nil, nil),
+    do: {:error, "Bad request. You must include at least one field to search on."}
+
+  def lookup(pid, last_name, first_name, date_of_birth) do
+    GenServer.call(pid, {:lookup, last_name, first_name, date_of_birth})
+  end
+
+  # Server
+
+  @impl GenServer
+  def init(opts) do
+    if Application.fetch_env!(:document_viewer, :live_catalog?) do
+      ets = :ets.new(__MODULE__, [:set, :protected])
+
+      {:ok, %__MODULE__{ets: ets, loaded?: false}, {:continue, {:load_documents, opts}}}
+    else
+      :ignore
+    end
+  end
+
+  @impl GenServer
+  @spec handle_continue({:load_documents, keyword()}, t()) :: {:noreply, t()}
+  def handle_continue({:load_documents, opts}, %__MODULE__{ets: ets, loaded?: false} = state) do
+    documents_fn = Keyword.get(opts, :documents_fn, &all_documents/0)
+
+    for document <- documents_fn.() do
+      _ =
+        :ets.insert(
+          ets,
+          {Document.key(document), document.last_name, document.first_name,
+           document.date_of_birth, document}
+        )
+    end
+
+    {:noreply, %__MODULE__{state | loaded?: true}}
```

Comment:
Do you need the `loaded?` boolean?

Reply:
It's not needed yet (so maybe it shouldn't be there yet), but my idea was to use it in a health check so the health check waits for the data to be loaded before reporting "ok".